pkg/kubelet/kubelet.go

```diff
@@ -2845,6 +2856,76 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
 	}, 100*time.Millisecond, stopCh)
 }
 
+// handleMaxContainerCheckpoints goes through the list of checkpoints of a
+// container in a pod and deletes old checkpoints if there are more than
+// 'kl.maxContainerCheckpoints'.
+func (kl *Kubelet) handleMaxContainerCheckpoints(podUID types.UID, podFullName, containerName string, options *runtimeapi.CheckpointContainerRequest) error {
+
+	if kl.maxContainerCheckpoints <= 0 {
+		// This cannot really happen as the kubelet config parser will
+		// error out early if this is <= 0. This additional error
+		// handling is mainly here to make sure tests are working.
+		return errors.New("maxContainerCheckpoints must be greater than 0")
+	}
+
+	// Lock while accessing the checkpoints on disk to avoid chaos
+	// if multiple threads would do it at the same time.
+	kl.maxContainerCheckpointsLock.Lock()
+	defer kl.maxContainerCheckpointsLock.Unlock()
+
+	checkpointDirectoryPath := kl.getCheckpointsDir()
+
+	checkpointDirectory, err := os.ReadDir(checkpointDirectoryPath)
+	if err != nil {
+		return fmt.Errorf("error reading checkpoint archive directory %s: %w", checkpointDirectoryPath, err)
+	}
+
+	regexString := fmt.Sprintf(`checkpoint-%s-%s-\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.*tar`, podFullName, containerName)
+
+	re, err := regexp.Compile(regexString)
+	if err != nil {
+		return fmt.Errorf("error compiling regex %s: %w", regexString, err)
+	}
+
+	var checkpointArchives = []fs.FileInfo{}
+	for _, archive := range checkpointDirectory {
+		// only select checkpoint archives which have been created from the same
+		// namespace/pod/container combination
+		if re.MatchString(archive.Name()) {
+			archiveFileinfo, err := archive.Info()
+			if err != nil {
+				return fmt.Errorf("error getting file info %s:%w", archive.Name(), err)
+			}
+			checkpointArchives = append(checkpointArchives, archiveFileinfo)
+		}
+	}
+
+	checkpointArchivesCounter := len(checkpointArchives)
+	if checkpointArchivesCounter <= int(kl.maxContainerCheckpoints) {
+		// Nothing to clean up, return early.
+		return nil
+	}
+
+	sort.Slice(checkpointArchives, func(i, j int) bool {
+		return checkpointArchives[i].ModTime().Before(checkpointArchives[j].ModTime())
```

Comment:
is ModTime stable and granular enough on all filesystems to always select the right checkpoints? should we bake the sort order we want into the filenames instead?

Reply:
> is ModTime stable and granular enough on all filesystems to always select the right checkpoints?

I think on ext4 and xfs I see very high timestamp resolution. Not sure with other file systems.

> should we bake the sort order we want into the filenames instead?

Sounds like a good idea. I will try to come up with something.

Reply:
No longer using the file's modification time for sorting but the new file name format which contains `UnixNano()` and a counter as suggested.